modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,174 @@
+<?php
+/**
+ * Optimizing for image loading optimization.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+require_once __DIR__ . '/class-ilo-html-tag-processor.php';
+
+/**
+ * Adds template output buffer filter for optimization if eligible.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+function ilo_maybe_add_template_output_buffer_filter() {
+	if ( ! ilo_can_optimize_response() ) {
+		return;
+	}
+	add_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' );
+}
+add_action( 'wp', 'ilo_maybe_add_template_output_buffer_filter' );
+
+/**
+ * Constructs preload links.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param array $lcp_images_by_minimum_viewport_widths LCP images keyed by minimum viewport width, amended with attributes key for the IMG attributes.
+ * @return string Markup for zero or more preload link tags.
+ */
+function ilo_construct_preload_links( array $lcp_images_by_minimum_viewport_widths ): string {
+	$preload_links = array();
+
+	$minimum_viewport_widths = array_keys( $lcp_images_by_minimum_viewport_widths );
+	for ( $i = 0, $len = count( $minimum_viewport_widths ); $i < $len; $i++ ) {
+		$lcp_element = $lcp_images_by_minimum_viewport_widths[ $minimum_viewport_widths[ $i ] ];
```

Comment:
I did it this way because for each minimum viewport width we also need to get the following one. So we're iterating over not just one viewport widths, but a window of two throughout the array.

Reply:
Ah, good point, I missed that. Can you maybe add a comment above this to clarify? For example:
`// This uses a for loop to be able to access the following element within the iteration, using a numeric index.`

Reply:
Done in 1d259b5a